platform/engineering/design-docs/2021-05-18-forms-library.md

```diff
@@ -0,0 +1,1025 @@
+
+### Logging
+N/A
+
+### Debugging
+
+#### Development logs
+Formulate will try to catch developer mistakes by emitting console logs in
+development builds similar to React. Mistakes it will attempt to catch include:
+- `FieldReview` components with no associated form data
+  - This is probably because the name (path to the data) is incorrect
+- Missing or invalid props
+
+#### Debugger object
+In development mode, the `Router` will attach a `Formulate` object to `window`.
+From the browser console, engineers will be able to call functions on
+`window.Formulate` to:
+- See all the form data
+- Set data for individual fields
+- Override the entire form data
+- See the list of routes
+
+### Caveats
+<!--
+Gotchas, differences between the design and implementation, other potential
+stumbling blocks for users or maintainers, and their implications and
+workarounds. Unless something is known to be tricky ahead of time, this section
+will probably start out empty.
+
+Rather than deleting it, it's recommended that you keep this section with a
+simple place holder, since caveats will almost certainly appear down the road.
+-->
+
+_To be determined._
+
+### Security Concerns
+N/A
+
+### Privacy Concerns
+N/A
+
+### Open Questions and Risks
+- Not using the same JSON schema to create the form fields as used in the data
+  validation opens the possibility that submitted data will fail validation
+  - For example, say the first name field doesn't have any validation in the
+    JSX, but the JSON schema contract with the API specifies a `maxLength`
+    of 25. A user can enter a name that's 27 characters long and not trigger any
+    validation errors until the submission step.
+  - This is a potential regression in UX
+  - To counteract this, the forms library will accept an optional JSON schema
+    and validate each form field against it.
+- Not using any kind of schema to create fields may lead to a field overwriting
+  bug
+  - If I give on input field the data path of `password` and another of
+    `password.confirm`, for example
+  - **Possible solution:** The answer here may be to somehow check all the
+    inputs against the form schema to ensure they're a match.
+
+### Work Estimates
+
+#### Milestone: Scaffold the project
+First, we have to prepare the library for development.
+
+**Prerequisites:** _None_
+
+**Tasks:**
+- Create the repo
+- Enable branch protection and code owners
+- Set up the TypeScript build process
+- Set up the testing tools
+- Set up `eslint` and `prettier`
+- Set up linting pre-commit hook
+
+**Estimate:** 1 day
+
+#### Milestone: Create the form builder fields
+This milestone is centered around getting the fields to look and feel like
+VA.gov fields.
+
+**Prerequisites:**
+- Scaffold the project
+
+**Tasks:**
+- Re-export Formik components such as `Formik` and `Form`
+- Build input fields
+  - Text box
+  - Text area
+  - Checkbox
+  - Radio group
+  - Select box
+  - File input
```

Comment:
Is this just referring to the Formik wrappers, or for the Web Components that they will be wrapping?

Reply:
Just the Formik wrappers. I didn't include the building the Web Components it'll use since that's work adjacent to the forms library itself.